index.js

```diff
@@ -180,2669 +180,2588 @@
+  supertypes, // :: Array Type
+  test,       // :: Array Type -> Any -> Boolean
+  tuples      // :: Array (Array3 String (a -> Array b) Type)
+) => globalThis.Object.assign (
+  globalThis.Object.create (Type$prototype, {
+    _extractors: {
+      value: tuples.reduce ((extractors, [k, e]) => ((
+        extractors[k] = e,
+        extractors
+      )), {}),
     },
-    'fantasy-land/equals': function(other) {
-      return (
-        Z.equals (this.type, other.type) &&
-        Z.equals (this.name, other.name) &&
-        Z.equals (this.url, other.url) &&
-        Z.equals (this.supertypes, other.supertypes) &&
-        this.keys.length === other.keys.length &&
-        this.keys.every (k => other.keys.includes (k)) &&
-        Z.equals (this.types, other.types)
-      );
+    _test: {
+      value: test,
+    },
+    extractors: {
+      value: tuples.reduce ((extractors, [k, e]) => ((
+        extractors[k] = x => toArray (e (x)),
+        extractors
+      )), {}),
+    },
+    format: {
+      value: format || ((outer, inner) =>
+        outer (name) +
+        Z.foldMap (
+          globalThis.String,
+          ([k, , t]) => (
+            t.arity > 0
+            ? outer (' ') + outer ('(') + inner (k) (show (t)) + outer (')')
+            : outer (' ')               + inner (k) (show (t))
+          ),
+          tuples
+        )
+      ),
     },
+  }),
+  {
+    arity,  // number of type parameters
+    keys: tuples.map (([k]) => k),
+    name,
+    supertypes,
+    type,
+    types: tuples.reduce ((types, [k, , t]) => ((types[k] = t, types)), {}),
+    url,
+  }
+);
+
+const BINARY        = 'BINARY';
+const FUNCTION      = 'FUNCTION';
+const INCONSISTENT  = 'INCONSISTENT';
+const NO_ARGUMENTS  = 'NO_ARGUMENTS';
+const NULLARY       = 'NULLARY';
+const RECORD        = 'RECORD';
+const UNARY         = 'UNARY';
+const UNKNOWN       = 'UNKNOWN';
+const VARIABLE      = 'VARIABLE';
+
+//  Inconsistent :: Type
+const Inconsistent = _Type (
+  INCONSISTENT,
+  '',
+  '',
+  0,
+  (outer, inner) => '???',
+  [],
+  null,
+  []
+);
+
+//  NoArguments :: Type
+const NoArguments = _Type (
+  NO_ARGUMENTS,
+  '',
+  '',
+  0,
+  (outer, inner) => '()',
+  [],
+  null,
+  []
+);
+
+//  functionUrl :: String -> String
+const functionUrl = name => {
+  const version = '0.22.0';  // updated programmatically
+  return (
+    `https://github.com/sanctuary-js/sanctuary-def/tree/v${version}#${name}`
+  );
+};
+
+const NullaryTypeWithUrl = name => supertypes => test => (
+  _NullaryType (name) (functionUrl (name)) (supertypes) (test)
+);
+
+const UnaryTypeWithUrl = name => supertypes => test => _1 => (
+  def (name)
+      ({})
+      ([Type, Type])
+      (_UnaryType (name) (functionUrl (name)) (supertypes) (test) (_1))
+);
+
+const BinaryTypeWithUrl = name => supertypes => test => _1 => _2 => (
+  def (name)
+      ({})
+      ([Type, Type, Type])
+      (_BinaryType (name) (functionUrl (name)) (supertypes) (test) (_1) (_2))
+);
```

Comment:
These functions now do more than before: they use `def` in addition to `functionUrl`.